xcm/pallet-xcm/src/lib.rs

```diff
@@ -1375,7 +1494,33 @@ pub mod pallet {
 					Self::deposit_event(Event::SupportedVersionChanged(origin, v));
 					0
 				},
-				(response, Some(QueryStatus::Pending { responder, maybe_notify, .. })) => {
+				(
+					response,
+					Some(QueryStatus::Pending {
+						responder, maybe_notify, maybe_match_querier, ..
+					}),
+				) => {
+					if let Some(match_querier) = maybe_match_querier {
+						let match_querier = match MultiLocation::try_from(match_querier) {
+							Ok(mq) => mq,
+							Err(_) => {
+								Self::deposit_event(Event::InvalidQuerierVersion(
+									origin.clone(),
+									query_id,
+								));
+								return 0
+							},
+						};
+						if querier.map_or(false, |q| q != &match_querier) {
```

Comment:
Hmm, this is somewhat surprising... in this code branch, I already have a `match_querier`, meaning that I do expect that there should be a querier location, but the `map_or(false, ..)` here means that I can refuse to supply a querier location, and it would still pass through, instead of being rejected.

@viniul Could you verify that my logic above is correct?